src/librustc/sir.rs

```diff
@@ -0,0 +1,136 @@
+    pub llvm_blocks: FxHashMap<*const BasicBlock, (SirFuncIdx, SirBlockIdx)>,
+    /// Function store. Also owns the blocks
+    pub funcs: IndexVec<SirFuncIdx, ykpack::Body>,
+}
+
+impl SirCx {
+    pub fn new() -> Self {
+        Self {
+            llvm_values: Default::default(),
+            llvm_blocks: FxHashMap::default(),
+            funcs: Default::default(),
+        }
+    }
+
+    pub fn add_func(&mut self, value: *const Value, symbol_name: String) {
+        let idx = SirFuncIdx::from_usize(self.funcs.len());
+
+        self.funcs.push(ykpack::Body{
+            symbol_name,
+            blocks: Default::default(),
+            num_locals: 0,  // FIXME
+            num_args: 0,    // FIXME
+            flags: 0,       // Set later.
+        });
+        let existing = self.llvm_values.insert(value, SirValue::Func(idx));
+        debug_assert!(existing.is_none());
+    }
+
+    pub fn add_block(&mut self, func: *const Value, block: *const BasicBlock) {
+        let func_idx = self.llvm_values[&func].func_idx();
+        let sir_func = &mut self.funcs[func_idx];
+        let block_idx = SirBlockIdx::from_usize(sir_func.blocks.len());
+        sir_func.blocks.push(ykpack::BasicBlock{
+            stmts: Default::default(),
+            term: ykpack::Terminator::Unreachable, // FIXME
+        });
+        let existing = self.llvm_blocks.insert(block, (func_idx, block_idx));
+        debug_assert!(existing.is_none());
+    }
+
+    pub fn get_symbol_name(&mut self, func: *const Value) -> &String {
+        let func_idx = self.llvm_values[&func].func_idx();
+        let sir_func = &mut self.funcs[func_idx];
+        &sir_func.symbol_name
+    }
+
+    /// For hardware tracing, during codegen we insert DILabels to know where we are in the binary.
+    /// These labels must be emitted in a deterministic order otherwise the reproducible build
+    /// checker gets upset. This function gives the codegen what it needs in a data structure which
+    /// can be iterated deterministically.
+    pub fn funcs_and_blocks_deterministic(&self)
+        -> IndexVec<SirFuncIdx, IndexVec<SirBlockIdx, *const BasicBlock>>
+    {
+        // Scaffold a data structure which can be iterated deterministically by the caller.
+        // To start, all pointer addresses are None. We would have use overwritable null pointers,
+        // but Rust won't let us and the pointee is not sized.
+        let mut res = IndexVec::from_elem_n(IndexVec::default(), self.funcs.len());
+        for (func_idx, func) in self.funcs.iter_enumerated() {
+            res[func_idx] = IndexVec::from_elem_n(None, func.blocks.len());
+        }
+
+        // Now populate the pointers.
+        for (bb, (func_idx, bb_idx)) in &self.llvm_blocks {
+            debug_assert!(res[*func_idx][*bb_idx] == None);
+            res[*func_idx][*bb_idx] = Some(*bb);
+        }
+
+        // Now get rid of the Option wrappers around our pointers. There should be no Nones left.
```

Comment:
The "There should be no `None`s left" seems redundant: `b.unwrap()` guarantees that, as does the type system given the return type for `ret`.

Reply:
eh? unwrap() doesn't guarantee anything. It can crash if there is a None.

You can think of this comment as one of your "unwrap() justifications" that you do in your PRs. If there was a None, then we did something wrong in the above loop.

Reply:
Ah you meant there should be no `None`s *before* the `unwrap`! The comment suggested *after* the `unwrap` to me. Maybe tweak the phrasing?